### ISSUE TITLE
Update Publish Provider Client

### ISSUE THREAD

**Issue by sjpadgett** [openemr/openemr#4043]
### Describe the problem

Publish is behind current APIs and authentication.

### Describe the solution you'd like
- Will need to update client to support OIDC authorization grant type and/or resource grant.
- This will allow queries to both outside and our APIs and server though, for FHIR only.

If someone wants to help here, please.....

**Comment by sjpadgett** [comment]
Have updated the oeHTTP client to handle OIDC OA2 authorization code grant authentication so we can have an internal client.
Works something like:(setDebug() is just a proxy port for exposing HTTP transactions. I use Fiddler)
```
$this->authClient = oeHttp::setDebug(8888)
            ->withOAuth($this->credentials, $this->authfhirBase . 'token', 'authorization_code')
            ->asFormParams();

        if ($this->authClient->token_storage->hasToken() === false && empty($code)) {
            $auth_url = 'https://opensourcedemr.us/oauth2/default/authorize?' . http_build_query([
                    'client_id' => $this->clientId,
                    'redirect_uri' => "https://opensourcedemr.us/phpfhir/providerPublishUI.php",
                    'response_type' => 'code',
                    'prompt' => 'select_account',
                    'scope' => implode(' ', [
                        'openid',
                        'api:fhir',
                        'api:pofh',
                        'email',
                    ]),
                    'state' => 'thisisme',
                ]);
            header("Location: $auth_url");
            exit;
        }
```
![image](https://user-images.githubusercontent.com/1859985/99760061-1425ac00-2ac1-11eb-9220-eb2dc8f5aafb.png)

Then tokens will persist and refresh automatically ie oeHTTP client will manage during api rest calls like:
```
$response = oeHttp::setDebug(8888)
                ->reAuth()
                ->usingBaseUri($this->fhirBase)->get($uri);
```
We're looking to have ability for Providers to hit outside FHIR servers or other resources. Or can stand up oeHTTP by itself.

I'm hoping someone will pick this project up and run with it.

**Comment by sjpadgett** [comment]
@bradymiller I have to cleanup/breakdown/rename the publish client support classes in phpfhir directory for PSR compatibility. I'm wondering if I should go ahead and add those classes to our OpenEMR namespace or just keep including the class files in scripts.

**Comment by bradymiller** [comment]
@sjpadgett , if not much more work, would get them into the src directory structure and OpenEMR namespace.

**Comment by sjpadgett** [comment]
Dear Abby,
Why do I voluntarily get myself involved with these interop mistresses?

**Comment by sjpadgett** [comment]
@bradymiller FYI
 The OAuth spec says that a refreshed access token can have the original scopes or fewer so must ensure the request doesn't include any new scopes. I may also do a rethink on client to user scope ties.

We'll need to decide on whether to force the sending of the id token by client. 
Currently, and though we support, I exclude basic authorization header for client_id and secret in discovery and state support of form post only. As my client uses basic and it is testing well, i'm going to include the basic authorization header as well. Of course, bearer is still used for access token.

I kept site:xxx scope but instead of extending the refresh grant to interrupt the flow, I inserted on the request interface side.
Working well and will put up PR after I do some code clean up.
